### ISSUE TITLE
`nx test` (jest) cannot resolve tsconfig paths after migration to nx 14.x

### ISSUE THREAD

**Issue by hevans90** [nrwl/nx#10660]
## Current Behavior

After migrating to `nx` 14, all my tests that import modules from aliased paths (from my libs) fail. The app builds fine and VSCode shows no issues with the module resolution, but when running tests, `jest` is no longer able to resolve my path aliases:

<img width="889" alt="image" src="https://user-images.githubusercontent.com/12142999/172862889-b93181d0-e15f-4d35-bd9d-e6b95210ca6f.png">

<img width="952" alt="image" src="https://user-images.githubusercontent.com/12142999/172862442-f5de0938-684f-4776-b13a-73f7b1f7fb26.png">

## Expected Behavior

The tests should pass.

## Steps to Reproduce

`nx migrate latest` on my `main` branch here:

https://github.com/hevans90/idleverse

Then `nx test` to run my default project's testing suite with the failing tests.

OR, check out this branch and `nx test`:

https://github.com/hevans90/idleverse/tree/build/nx-14

___

This issue may not be prioritized if details are not provided to help us reproduce the issue.

### Environment
   Node : 16.13.2
   OS   : darwin x64
   yarn : 1.22.18

   nx : 14.2.2
   @nrwl/angular : Not Found
   @nrwl/cypress : 14.2.2
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.2.2
   @nrwl/eslint-plugin-nx : 14.2.2
   @nrwl/express : Not Found
   @nrwl/jest : 14.2.2
   @nrwl/js : 14.2.2
   @nrwl/linter : 14.2.2
   @nrwl/nest : Not Found
   @nrwl/next : Not Found
   @nrwl/node : 14.2.2
   @nrwl/nx-cloud : 14.0.8
   @nrwl/nx-plugin : Not Found
   @nrwl/react : 14.2.2
   @nrwl/react-native : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : 14.2.2
   @nrwl/web : 14.2.2
   @nrwl/workspace : 14.2.2
   typescript : 4.7.3

**Comment by barbados-clemens** [comment]
Hi @hevans90, I looked into this and it's pretty odd what's happening. I don't see anything that stands out as strange with your setup, but if I had to venture to guess it's something config-related. I'll try to take a look later this week.

the _only_ thing that I see is different is the using jest.config.js files instead of jest.config.ts files, which changed in 14.1.5 I believe. Though using a js file is completely fine and should continue to work. 

Maybe try giving ts-jest a try to see if you get any different behavior. I've had strange behavior from babel-jest before and using ts-jest usually shows what extras I might need to provide to babel-jest. 

Another thing to try is to remove any extra customizations. just default jest configs and default libraries and see if you're still seeing this behavior and add back 1 thing at a time to see what might be causing the issue.

**Comment by hevans90** [comment]
@barbados-clemens funny you mention the `jest.config.ts` as I just spun up a new `nx 14.2.x` workspace and was trying to compare setups and flagged this. Sadly (as you mentioned) changing this has no effect.

I'll try running ts-jest as see what happens!

**Comment by jonwalstedt** [comment]
Any luck on this one? I am getting the same errors

**Comment by jonwalstedt** [comment]
It starts failing when updating from @nrwl/jest 14.1.4 to 14.1.5

Somewhere here I guess:
git diff ac76a7e53abb6a7580441dada9fb0ec2aab82035 5251bd73acb5ac52d37b0795e26346b02dba24f7 -- packages/jest > changes.diff

**Comment by hevans90** [comment]
@jonwalstedt thanks a lot for a workaround. Can confirm that downgrading `@nrwl/jest` from `14.2.2` to `14.1.14` seems to allow the module resolver to work with `tsconfig.paths`.

Regarding `ts-jest` I get the same errors when running `nx test` for this project:

<img width="946" alt="image" src="https://user-images.githubusercontent.com/12142999/172939934-724f8242-b8fa-4335-a8ff-a5b4394054ea.png">

**Comment by barbados-clemens** [comment]
@jonwalstedt thanks for the diff, it looks like it was when I changed the nxPreset. Looking at your root jest.preset.js it doesn't have the updated version @hevans90 

You'll want to add `.default` to the nxPreset import. 
```diff
- const nxPreset = require('@nrwl/jest/preset').default;
+ const nxPreset = require('@nrwl/jest/preset').default;

module.exports = { ...nxPreset };
```

This should have been fixed in v14.1.5 migration for you. You can try rerunning that migration to see if it's added as you're preset is the default so there isn't a reason it shouldn't work. Or just manually make that change. This was causing the @nrwl/jest/plugins/resolver to not resolve the module names.

**Comment by jonwalstedt** [comment]
That did the trick, thanks @barbados-clemens

**Comment by hevans90** [comment]
Thanks a lot @barbados-clemens - worked!

**Comment by christophechevalier** [comment]
> @jonwalstedt thanks for the diff, it looks like it was when I changed the nxPreset. Looking at your root jest.preset.js it doesn't have the updated version @hevans90
> 
> You'll want to add `.default` to the nxPreset import.
> 
> ```diff
> - const nxPreset = require('@nrwl/jest/preset');
> + const nxPreset = require('@nrwl/jest/preset').default;
> 
> module.exports = { ...nxPreset };
> ```
> 
> This should have been fixed in v14.1.5 migration for you. You can try rerunning that migration to see if it's added as you're preset is the default so there isn't a reason it shouldn't work. Or just manually make that change. This was causing the @nrwl/jest/plugins/resolver to not resolve the module names.

@barbados-clemens Without this workaround, when I run `yarn jest MyProject`, it looks likes alias paths are not recognized, true. But it seems not works completly in my project or rather, it generates errors for me on imports into files that have nothing to do with it.  

### Environment  
![image](https://user-images.githubusercontent.com/7148846/190460122-08ff0cce-5081-4ce3-9fa4-33c52d28f0bf.png)


**jest.config.base.js**  
```js
module.exports = {
  testEnvironment: 'jsdom',
  transform: {
    '^.+\\.(ts|js|mjs|html|svg)$': 'jest-preset-angular',
  },
  setupFilesAfterEnv: ['<rootDir>/src/jest.ts'],
  transformIgnorePatterns: [`./node_modules/(?!.*\\.mjs$)`],
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment',
  ],
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
      stringifyContentPathRegex: '\\.(html|svg)$',
    },
  },
  globalSetup: 'jest-preset-angular/global-setup',
};
```
**jest.ts**
```ts
// Need to be imported in first position for async() functionality
import 'jest-preset-angular/setup-jest';

// Other imports
import '@angular-extensions/pretty-html-log';
```
**jest.config.js** example for each client project  
```js
const jestBaseConfig = require('../../jest.config.base');

module.exports = {
  ...jestBaseConfig,
  displayName: 'my-project',
  preset: '../../jest.preset.js',
  coverageDirectory: '../../coverage/apps/my-project',
  testPathIgnorePatterns: ['<rootDir>/src/tests'],
};
```
Here, some screenshots of the failed tests resume.   
![image](https://user-images.githubusercontent.com/7148846/190455330-f05e4455-dd6f-4b7d-accb-f67631071aad.png)  
![image](https://user-images.githubusercontent.com/7148846/190456967-2515ea63-5e33-4f4b-ab7d-2ba0e1660377.png)   
![image](https://user-images.githubusercontent.com/7148846/190458351-6c502860-239e-4b7f-8db4-bf7722d966b0.png)    

ReferenceError is not logic because all failed tests does not imported or used this selector or interface.  
I don't know if it's related at `@nrwl/jest/plugins/resolver`, but this random error appears in others tests files too.  

I will try to reproduce that in [Stackblitz](https://stackblitz.com).  
If you have any suggestions, I'll take it. Thanks

**Comment by barbados-clemens** [comment]
is this with jest 28 @christophechevalier?
this looks an error @FrozenPandaz ran into, which is basically ts-jest/jest-preset-angular was returning "different" transpiled code causing a null ref error since an import wasn't being included. I think we are still trying to debug this ourselves, but doesn't look to be directly an nx issue but maybe a transformer/downstream dep. 

but any reproduceable would be helpful. we haven't had much luck trying to track it down exactly where it's happening.

**Comment by christophechevalier** [comment]
Jest 28 yes.  
```json
"devDependencies": {
  "jest": "28.1.0",
  "jest-environment-jsdom": "28.1.0",
  "jest-preset-angular": "12.2.2",
  "ts-jest": "28.0.3",
}
```  

I tried to upgrade the version of Jest to use 29.0.3 but I have compatibility issues with `ts-jest` which does not accept a version greater than or equal to 29.0.0. So I kept my current versions.  

Stay tuned please, we might ask you again to review the code that reproduces our issue on [Stackblitz](https://stackblitz.com/) soon.

**Comment by jcabannes** [comment]
Hello @barbados-clemens , I work with @christophechevalier . I didn't succeed to make a Stackblitz project reproducing the error, so I created a small project on github, available at this [address](https://github.com/jcabannes/nx-angular-ngxs-jest). If you have any questions or problems do not hesitate to contact us.

**Comment by barbados-clemens** [comment]
thanks for the reproduce-able @jcabannes and @christophechevalier

I was able to get the tests passing by setting `    `emitDecoratorMetadata": false` in the project level `tsconfig.spec.json` where the error is happening. 

Assuming you don't need reflection at test time this should be find to do. If you do need reflection at test time, then this _could_ be an issue with angular devkit as [noted in this issue from ngxs](https://github.com/ngxs/store/issues/1720)

Here is a PR against your reproduce-able showing it working.
https://github.com/jcabannes/nx-angular-ngxs-jest/pull/1

**Comment by jcabannes** [comment]
Thanks for your PR @barbados-clemens , this solves the problem :)

**Comment by gabrielgagne** [comment]
Had the same issue here with the decorators after upgrading nx anywhere from 14.6.5 and up. 
The emitDecoratorMetadata: false seems to fix it, although I have yet to test it in a couple more dozen libs.

**Comment by jbjhjm** [comment]
Guys I managed to write an isolated test and I think I found some interesting things...

deptest-imports.ts
```ts
export interface DepTestInterface {
	endpoint?:string,
	userId?:number, 
	permissions?:string[],
}

export class DepTestClass {
	endpoint?:string
	userId?:number
	permissions?:string[]
}
```

deptest-interface.ts
```ts
export interface InterfaceFromInterfaceOnlyModule {
	endpoint?:string,
	userId?:number, 
	permissions?:string[],
}
export const DUMMY = null;
```

deptest.spec.ts
```ts

import { DepTestClass, DepTestInterface } from './deptest-imports';
import { InterfaceFromInterfaceOnlyModule , DUMMY } from './deptest-interface';

export function MyDecorator():MethodDecorator {
	return <T>(target,propertyKey,descriptor:TypedPropertyDescriptor<T>) => {}
}

export interface DepTestInlineInterface {
	endpoint?:string, // endpoint used to fetch user data from
	userId?:number, // type defines 
	permissions?:string[],
}

class X {
	// All fine ->
	static noDecorator1(state:any):DepTestInterface {return state.tokenData||null}
	static noDecorator2(state:any):InterfaceFromInterfaceOnlyModule {return state.tokenData||null}
	@MyDecorator()	static test2(state:any):number {return state.tokenData?.userId||null}
	@MyDecorator()	static test3(state:any):DepTestInlineInterface {return state.tokenData||null}
	// from imported module, still fine! ->
	@MyDecorator()	static test4(state:any):DepTestClass {return state.tokenData||null}
	@MyDecorator()	static test5(state:any):DepTestInterface {return state.tokenData||null}
	// errors ->
	@MyDecorator()	static test6(state:any):InterfaceFromInterfaceOnlyModule {return state.tokenData||null}
}

describe('test', () => {
	it('should work', () => { expect(true).toBe(true) });
});

```

So far, executing that test with jest 28 + ts-node fails reliably with same error:
`ReferenceError: deptest_interface_1 is not defined`

I'm concluding following circumstances lead to the error:

- Import a interface from a module that contains nothing but interfaces 
- Use it on a method that is being decorated
- Have emitDecoratorMetadata enabled

Interestingly, only the last static function declaration errors, all of those before are fine.
So this means it will only happen if a module exports only types. As soon as something else is exported and the module actually "exists", the error is gone.
You'll see the test contains an unused "DUMMY" import. As soon as this is being used anywhere, the ReferenceError will disappear because the imported module is not type-only anymore.

Summing up, this is not a nx issue.
Question is... where does this issue happen?
